Python/Product/PythonTools/PythonTools/Extensions.cs

```diff
@@ -255,23 +255,113 @@ public static class Extensions {
+        /// 
+        /// For files on disk this is pretty easy - we analyze each file on it's own in a buffer parser.  
+        /// Therefore we map filename -> analyzer and then get the analysis from teh analyzer.  If we
+        /// determine an analyzer but the file isn't loaded into it for some reason this would return null.
+        /// We can also apply some policy to buffers depending upon the view that they're hosted in.  For
+        /// example if a buffer is outside of any projects, but hosted in a difference view with a buffer
+        /// that is in a project, then we'll look in the view that has the project.
+        /// 
+        /// For interactive windows we will use the analyzer that's configured for the window.
+        /// </summary>
+        /// <param name="view"></param>
+        /// <param name="buffer"></param>
+        /// <param name="provider"></param>
+        /// <param name="entry"></param>
+        /// <returns></returns>
+        internal static bool TryGetAnalysisEntry(this ITextView view, ITextBuffer buffer, IServiceProvider provider, out AnalysisEntry entry) {
+            PythonReplEvaluator evaluator;
+            IInteractiveEvaluator interactiveEval;
+            if ((buffer.Properties.TryGetProperty(typeof(IInteractiveEvaluator), out interactiveEval) 
+                && (evaluator = interactiveEval as PythonReplEvaluator) != null) ||
+                (view != null && view.Properties.TryGetProperty(typeof(PythonReplEvaluator), out evaluator))) {
+
+                var analyzer = evaluator.ReplAnalyzer;
+                entry = analyzer.GetAnalysisEntryFromPath(evaluator.AnalysisFilename);
+                return entry != null;
             }
+
+            string path = buffer.GetFilePath();
+            if (path != null) {
+                var docTable = provider.GetService(typeof(SVsRunningDocumentTable)) as IVsRunningDocumentTable4;
+                var cookie = docTable.GetDocumentCookie(path);
+                VsProjectAnalyzer analyzer = null;
+                if (cookie != VSConstants.VSCOOKIE_NIL) {
+                    IVsHierarchy hierarchy;
+                    uint itemid;
+                    docTable.GetDocumentHierarchyItem(cookie, out hierarchy, out itemid);
+                    if (hierarchy != null) {
+                        var pyProject = hierarchy.GetProject()?.GetPythonProject();
+                        if (pyProject != null) {
+                            analyzer = pyProject.GetAnalyzer();
+                        }
+                    }
+                }
+
+                if (analyzer == null && view != null) {
+                    // We could spin up a new analyzer for non Python projects...
+                    analyzer = view.GetBestAnalyzer(provider);
+                }
+
+                if (analyzer != null) {
+                    entry = analyzer.GetAnalysisEntryFromPath(path);
+                    return entry != null;
```

Comment:
This is a good place to use `entry = analyzer?.GetAnalysisEntryFromPath(path)` and reduce some control flow.